src/main/java/org/openrewrite/staticanalysis/RemoveToStringCallsFromArrayInstances.java

```diff
@@ -0,0 +1,116 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.staticanalysis;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Preconditions;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.UsesMethod;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.JavaType;
+
+import java.util.List;
+
+public class RemoveToStringCallsFromArrayInstances extends Recipe {
+    private static final MethodMatcher TO_STRING_MATCHER = new MethodMatcher("java.lang.Object toString(..)");
+    private static final MethodMatcher PRINTLN_MATCHER = new MethodMatcher("java.io.PrintStream println(..)");
+    private static final MethodMatcher STR_FORMAT_MATCHER = new MethodMatcher("java.lang.String format(..)");
+
+    @Override
+    public String getDisplayName() {
+        return "toString should not be called on array instances";
+    }
+
+    @Override
+    public String getDescription() {
+        return "toString should not be called on array instances.";
+    }
+
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(Preconditions.or(
+                new UsesMethod<>(TO_STRING_MATCHER),
+                new UsesMethod<>(PRINTLN_MATCHER),
+                new UsesMethod<>(STR_FORMAT_MATCHER)
+        ), new RemoveToStringFromArraysVisitor());
+    }
+
+    private static class RemoveToStringFromArraysVisitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
+            J.MethodInvocation m = super.visitMethodInvocation(mi, ctx);
+
+            if (TO_STRING_MATCHER.matches(m)) {
+                String builder_string = "Arrays.toString(#{anyArray(java.lang.String)})";
+                Expression select = m.getSelect();
+                assert select != null;
+
+                return buildReplacement(builder_string, m, select);
+            }else if (PRINTLN_MATCHER.matches(m)) {
+                Expression select = m.getArguments().get(0);
+                String builder_string = "System.out.println(Arrays.toString(#{anyArray(java.lang.String)}))";
+
+                return buildReplacement(builder_string, m, select);
+            }else if (STR_FORMAT_MATCHER.matches(m)) {
+                List<Expression> arguments = m.getArguments();
+                StringBuilder builder_string = new StringBuilder("String.format(#{any(java.lang.String)}");
+                boolean arrayExists = false;
+
+                for (int i = 0; i<arguments.size(); i++) {
```

Comment:
For reasons of efficiency (and simplicity in a way) I am wondering if we shouldn't instead use cursor messaging and transform this bottom-up. What I mean is that you in this method would add a message to the cursor to indicate that you have found one of the methods which need transforming. After that you would do the `super` call and then you would add an override for `visitExpression()` which in case the parent cursor (skipping over any intermediate `J.Parentheses` cursors) has the mentioned message set and the expression is array-typed would return the array wrapped in `Arrays#toString()`. Does that make sense to you? This logic would then be exactly the same for both `STR_FORMAT_MATCHER` and `PRINTLN_MATCHER` and all the other methods I mentioned.

Reply:
Yep this makes sense. I'll get started on this change.